rpc/gzip.go

```diff
@@ -0,0 +1,50 @@
+package rpc
+
+import (
+	"compress/gzip"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"sync"
+)
+
+var gzPool = sync.Pool{
+	New: func() interface{} {
+		w := gzip.NewWriter(ioutil.Discard)
+		return w
+	},
+}
+
+type gzipResponseWriter struct {
+	io.Writer
+	http.ResponseWriter
+}
+
+func (w *gzipResponseWriter) WriteHeader(status int) {
+	w.Header().Del("Content-Length")
+	w.ResponseWriter.WriteHeader(status)
+}
+
+func (w *gzipResponseWriter) Write(b []byte) (int, error) {
+	return w.Writer.Write(b)
+}
+
+func Gzip(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if !strings.Contains(r.Header.Get("Accept-Encoding"), "gzip") {
+			next.ServeHTTP(w, r)
+			return
+		}
+
+		w.Header().Set("Content-Encoding", "gzip")
+
+		gz := gzPool.Get().(*gzip.Writer)
+		defer gzPool.Put(gz)
+
+		gz.Reset(w)
+		defer gz.Close()
```

Comment:
I think this should be `gz.Flush()`. Looking at the code of `gzip`, Close also sets the state of the writer to `closed`, which is not reset on Reset.

Reply:
The `Close` is required to write the gzip footer. `gz.Reset` does take care of resetting `closed` by replacing the `Writer` while keeping the compressor (through `init`).
Looking at https://github.com/nytimes/gziphandler/blob/master/gzip.go#L332 it's a pretty common pattern so I think this is actually correct